packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m

```diff
@@ -0,0 +1,220 @@
+    _remoteLoadTimeout = remoteLoadTimeout;
+    _encounteredErrors = [NSMutableArray new];
+  }
+  return self;
+}
+
+- (void)handleError:(NSError *)error fromLaunchedUpdate:(nullable EXUpdatesUpdate *)launchedUpdate
+{
+  if (launchedUpdate) {
+    [self _markUpdateFailed:launchedUpdate];
+  }
+  [self _startPipelineWithEncounteredError:error];
+}
+
+- (void)handleException:(NSException *)exception fromLaunchedUpdate:(nullable EXUpdatesUpdate *)launchedUpdate
+{
+  if (launchedUpdate) {
+    [self _markUpdateFailed:launchedUpdate];
+  }
+  [self _startPipelineWithEncounteredError:exception];
+}
+
+- (void)notifyNewRemoteLoadStatus:(EXUpdatesRemoteLoadStatus)newStatus
+{
+  dispatch_async(_errorRecoveryQueue, ^{
+    if (!self->_isWaitingForRemoteUpdate) {
+      return;
+    }
+    self->_isWaitingForRemoteUpdate = NO;
+    if (newStatus == EXUpdatesRemoteLoadStatusNewUpdateLoaded) {
+      [self->_pipeline insertObject:@(EXUpdatesErrorRecoveryTaskLaunchNew) atIndex:0];
+    }
+    [self _runNextTask];
+  });
+}
+
+# pragma mark - pipeline tasks
+
+- (void)_startPipelineWithEncounteredError:(id)encounteredError
+{
+  dispatch_async(_errorRecoveryQueue, ^{
+    [self->_encounteredErrors addObject:encounteredError];
+    if (!self->_isRunning) {
+      self->_isRunning = YES;
+      [self _runNextTask];
+    }
+  });
+}
+
+- (void)_runNextTask
+{
+  dispatch_assert_queue(_errorRecoveryQueue);
+  NSNumber *nextTask = _pipeline[0];
+  [_pipeline removeObjectAtIndex:0];
+  switch ((EXUpdatesErrorRecoveryTask)nextTask.integerValue) {
+    case EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate:
+      [self _waitForRemoteLoaderToFinish];
+      break;
+    case EXUpdatesErrorRecoveryTaskLaunchNew:
+    case EXUpdatesErrorRecoveryTaskLaunchCached:
+      [self _tryRelaunchFromCache];
+      break;
+    case EXUpdatesErrorRecoveryTaskLaunchEmbedded:
+      [self _tryLaunchEmbeddedUpdate];
+      break;
+    case EXUpdatesErrorRecoveryTaskCrash:
+      [self _crash];
+      break;
+    default:
+      @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"EXUpdatesErrorRecovery cannot run the next task" userInfo:@{@"nextTaskValue": nextTask}];
+  }
+}
+
+- (void)_markUpdateFailed:(EXUpdatesUpdate *)update
+{
+  dispatch_async(_delegate.database.databaseQueue, ^{
+    NSError *error;
+    [self.delegate.database markUpdateFailed:update error:&error];
+    if (error) {
+      NSLog(@"Unable to mark update as failed in the local DB: %@", error.localizedDescription);
+      dispatch_async(self->_errorRecoveryQueue, ^{
+        [self->_pipeline removeObject:@(EXUpdatesErrorRecoveryTaskLaunchCached)];
+      });
+    }
+  });
+}
+
+- (void)_waitForRemoteLoaderToFinish
+{
+  dispatch_assert_queue(_errorRecoveryQueue);
+  if (_delegate.remoteLoadStatus == EXUpdatesRemoteLoadStatusLoading) {
+    _isWaitingForRemoteUpdate = YES;
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, _remoteLoadTimeout * NSEC_PER_MSEC), _errorRecoveryQueue, ^{
```

Comment:
I left a comment [here](https://github.com/expo/expo/pull/14398#pullrequestreview-753467235) explaining why I am uncomfortable with using timer instead of subscribing specific events to trigger different phases of error recovery. This way we can reduce uncertainty in scenarios where the app is already in an uncertain state due to an error.